iroh-sync/src/sync.rs

```diff
@@ -733,16 +798,54 @@ pub struct Record {
     timestamp: u64,
 }
 
+impl RangeValue for Record {}
+
+/// Ordering for entry values.
+///
+/// Compares first the timestamp, then the content hash.
+impl Ord for Record {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.timestamp
+            .cmp(&other.timestamp)
+            .then_with(|| self.hash.cmp(&other.hash))
+    }
+}
+
+impl PartialOrd for Record {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
 impl Record {
     /// Create a new record.
     pub fn new(hash: Hash, len: u64, timestamp: u64) -> Self {
+        debug_assert!(
+            len != 0 || hash == Hash::EMPTY,
+            "if `len` is 0 then `hash` must be the hash of the empty byte range"
+        );
         Record {
             hash,
             len,
             timestamp,
         }
     }
 
+    /// Create a tombstone record (empty content)
+    pub fn empty(timestamp: u64) -> Self {
+        Self::new(Hash::EMPTY, 0, timestamp)
+    }
+
+    /// Create a tombstone record with the timestamp set to now.
+    pub fn empty_current() -> Self {
+        Self::new_current(Hash::EMPTY, 0)
+    }
+
+    /// Return `true` if the entry is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len == 0 && self.hash == Hash::EMPTY
```

Comment:
we use || in other places, maybe here too?

Reply:
we should probably do a consistency check at some point, that if the hash is `EMPTY` the length is set to `0`

Reply:
This is mostly covered already:
* `replica.insert()` with `hash::EMPTY` or `len: 0` returns an error so that you're forced to use `wipe_at_prefix` That is debatable - should we allow it instead?
* `wipe_at_prefix` sets `len: 0` and `hash::EMPTY`
* leaves `insert_remote_entry` for gossiped entries - I'll add a check there and return an error for entries with hash::Empty and not len 0.

Reply:
Added the validation on `insert_remote_entry`